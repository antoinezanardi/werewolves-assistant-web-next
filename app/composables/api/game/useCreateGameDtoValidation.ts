import { storeToRefs } from "pinia";
import { group as groupByField } from "radash";
import type { ComputedRef, Ref } from "vue";

import { MIN_PLAYERS_IN_GAME, MIN_PLAYERS_IN_GROUP } from "~/composables/api/game/constants/game.constants";
import type { CreateGamePlayerWithGroupDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { Role } from "~/composables/api/role/types/role.class";
import { useRolesStore } from "~/stores/role/useRolesStore";

type RoleWithMinInGame = Role & { minInGame: number };

type UseCreateGameDtoValidation = {
  isMinimumPlayersReached: ComputedRef<boolean>;
  areAllRolesSet: ComputedRef<boolean>;
  doesContainOneVillagerSidedRole: ComputedRef<boolean>;
  doesContainOneWerewolfSidedRole: ComputedRef<boolean>;
  areRolesMinimumPlayersReached: ComputedRef<boolean>;
  isTwoSistersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  isThreeBrothersRolePresentAndMinimumPlayersReached: ComputedRef<boolean>;
  areAdditionalCardsSetForThiefIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForActorIfPresent: ComputedRef<boolean>;
  areAdditionalCardsSetForAdditionalCardsDependantRoles: ComputedRef<boolean>;
  arePlayerGroupsSetForPrejudicedManipulatorIfPresent: ComputedRef<boolean>;
  canCreateGame: ComputedRef<boolean>;
  gameCreationValidationErrors: ComputedRef<string[]>;
  isRolePresentAndMinimumPlayersReached: (role: RoleWithMinInGame) => boolean;
};

function useCreateGameDtoValidation(createGameDto: Ref<CreateGameDto>): UseCreateGameDtoValidation {
  const rolesStore = useRolesStore();
  const { roles } = storeToRefs(rolesStore);

  const { t } = useI18n();

  const isMinimumPlayersReached = computed<boolean>(() => createGameDto.value.players.length >= MIN_PLAYERS_IN_GAME);

  const areAllRolesSet = computed<boolean>(() => createGameDto.value.players.every(player => player.role.name !== undefined));

  const doesContainOneVillagerSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "villagers"));

  const doesContainOneWerewolfSidedRole = computed<boolean>(() => createGameDto.value.players.some(player => player.side.current === "werewolves"));

  const areRolesMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const rolesWithMinimumPlayers = roles.value.filter((role): role is RoleWithMinInGame => role.minInGame !== undefined);

    return rolesWithMinimumPlayers.every(role => isRolePresentAndMinimumPlayersReached(role));
  });

  const isTwoSistersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "two-sisters") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const isThreeBrothersRolePresentAndMinimumPlayersReached = computed<boolean>(() => {
    if (!roles.value) {
      return false;
    }
    const role = roles.value.find(({ name }) => name === "three-brothers") as RoleWithMinInGame;

    return isRolePresentAndMinimumPlayersReached(role);
  });

  const areAdditionalCardsSetForThiefIfPresent = computed<boolean>(() => {
    const isThiefPresent = createGameDto.value.players.some(player => player.role.name === "thief");
    if (!isThiefPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "thief") === true;
  });

  const areAdditionalCardsSetForActorIfPresent = computed<boolean>(() => {
    const isActorPresent = createGameDto.value.players.some(player => player.role.name === "actor");
    if (!isActorPresent) {
      return true;
    }
    return createGameDto.value.additionalCards?.some(card => card.recipient === "actor") === true;
  });

  const areAdditionalCardsSetForAdditionalCardsDependantRoles = computed<boolean>(() => areAdditionalCardsSetForActorIfPresent.value &&
    areAdditionalCardsSetForThiefIfPresent.value);

  const arePlayerGroupsSetForPrejudicedManipulatorIfPresent = computed<boolean>(() => {
    const isPrejudicedManipulatorPresent = createGameDto.value.players.some(player => player.role.name === "prejudiced-manipulator");
    if (!isPrejudicedManipulatorPresent) {
      return true;
    }
    const minimumGroups = 2;
    const playersWithGroup = createGameDto.value.players.filter(player => player.group !== undefined) as CreateGamePlayerWithGroupDto[];
    const groups = groupByField(playersWithGroup, player => player.group) as Record<string, CreateGamePlayerWithGroupDto[]>;
    const areAllGroupsMinimumPlayersReached = Object.values(groups).every(group => group.length >= MIN_PLAYERS_IN_GROUP);

    return Object.keys(groups).length >= minimumGroups && areAllGroupsMinimumPlayersReached;
  });

  const canCreateGame = computed<boolean>(() => isMinimumPlayersReached.value &&
    areAllRolesSet.value &&
    doesContainOneVillagerSidedRole.value &&
    doesContainOneWerewolfSidedRole.value &&
    areRolesMinimumPlayersReached.value &&
    areAdditionalCardsSetForThiefIfPresent.value &&
    areAdditionalCardsSetForActorIfPresent.value &&
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value);

  const gameCreationValidationErrors = computed<string[]>(() => {
    const errors: Record<string, boolean> = {
      [t("composables.useCreateGameDtoValidation.minimumOfPlayersNotReached")]: !isMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.notAllRolesAreSet")]: !areAllRolesSet.value,
      [t("composables.useCreateGameDtoValidation.noVillagerSidedRole")]: !doesContainOneVillagerSidedRole.value,
      [t("composables.useCreateGameDtoValidation.noWerewolfSidedRole")]: !doesContainOneWerewolfSidedRole.value,
      [t("composables.useCreateGameDtoValidation.twoSistersMinimumPlayersNotReached")]: !isTwoSistersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.threeBrothersMinimumPlayersNotReached")]: !isThreeBrothersRolePresentAndMinimumPlayersReached.value,
      [t("composables.useCreateGameDtoValidation.thiefAdditionalCardsNotSet")]: !areAdditionalCardsSetForThiefIfPresent.value,
      [t("composables.useCreateGameDtoValidation.actorAdditionalCardsNotSet")]: !areAdditionalCardsSetForActorIfPresent.value,
      [t("composables.useCreateGameDtoValidation.prejudicedManipulatorGroupsNotSet")]: !arePlayerGroupsSetForPrejudicedManipulatorIfPresent.value,
    };

    return Object.keys(errors).filter((key: keyof typeof errors) => errors[key]);
  });

  function isRolePresentAndMinimumPlayersReached(role: RoleWithMinInGame): boolean {
    const playersWithRole = createGameDto.value.players.filter(player => player.role.name === role.name);
    const playersWithRoleCount = playersWithRole.length;

    return playersWithRoleCount === 0 || playersWithRoleCount >= role.minInGame;
  }
  return {
    isMinimumPlayersReached,
    areAllRolesSet,
    doesContainOneVillagerSidedRole,
    doesContainOneWerewolfSidedRole,
    areRolesMinimumPlayersReached,
    isTwoSistersRolePresentAndMinimumPlayersReached,
    isThreeBrothersRolePresentAndMinimumPlayersReached,
    areAdditionalCardsSetForThiefIfPresent,
    areAdditionalCardsSetForActorIfPresent,
    areAdditionalCardsSetForAdditionalCardsDependantRoles,
    arePlayerGroupsSetForPrejudicedManipulatorIfPresent,
    canCreateGame,
    gameCreationValidationErrors,
    isRolePresentAndMinimumPlayersReached,
  };
}

export { useCreateGameDtoValidation };