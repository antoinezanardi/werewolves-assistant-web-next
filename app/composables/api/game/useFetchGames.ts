import type { CreateGameFeedbackDto } from "~/composables/api/game/dto/create-game-feedback/create-game-feedback.dto";
import type { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { Game } from "~/composables/api/game/types/game.class";
import { useWerewolvesAssistantApi } from "~/composables/api/useWerewolvesAssistantApi";

type UseFetchGames = {
  createGame: (createGameDto: CreateGameDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  getGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  cancelGame: (gameId: string) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  makeGamePlay: (gameId: string, makeGamePlayDto: MakeGamePlayDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
  createGameFeedback: (gameId: string, createGameFeedbackDto: CreateGameFeedbackDto) => Promise<ReturnType<typeof $fetch<Game>> | null>;
};

function useFetchGames(): UseFetchGames {
  const { fetchWerewolvesAssistantApi } = useWerewolvesAssistantApi();

  async function createGame(createGameDto: CreateGameDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games`, {
        method: "POST",
        body: JSON.stringify(createGameDto),
      });
    } catch {
      return null;
    }
  }

  async function getGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "GET" });
    } catch {
      return null;
    }
  }

  async function cancelGame(gameId: string): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}`, { method: "DELETE" });
    } catch {
      return null;
    }
  }

  async function makeGamePlay(gameId: string, makeGamePlayDto: MakeGamePlayDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/play`, {
        method: "POST",
        body: JSON.stringify(makeGamePlayDto),
      });
    } catch {
      return null;
    }
  }

  async function createGameFeedback(gameId: string, createGameFeedbackDto: CreateGameFeedbackDto): Promise<ReturnType<typeof $fetch<Game>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Game>(`/games/${gameId}/feedback`, {
        method: "POST",
        body: JSON.stringify(createGameFeedbackDto),
      });
    } catch {
      return null;
    }
  }
  return {
    createGame,
    getGame,
    cancelGame,
    makeGamePlay,
    createGameFeedback,
  };
}

export { useFetchGames };