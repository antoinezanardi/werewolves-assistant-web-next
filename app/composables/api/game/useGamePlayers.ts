import type { ComputedRef, Ref } from "vue";

import type { Game } from "~/composables/api/game/types/game.class";
import type { Player } from "~/composables/api/game/types/players/player.class";
import type { RoleName } from "~/composables/api/role/types/role.types";

type UseGamePlayers = {
  sheriffInPlayers: ComputedRef<Player | undefined>;
  getPlayersWithCurrentRole: (role: RoleName) => Player[];
};

function useGamePlayers(game: Ref<Game>): UseGamePlayers {
  const sheriffInPlayers = computed<Player | undefined>(() => game.value.players.find(({ attributes }) => attributes.some(({ name }) => name === "sheriff")));

  function getPlayersWithCurrentRole(role: RoleName): Player[] {
    return game.value.players.filter(player => player.role.current === role);
  }
  return {
    getPlayersWithCurrentRole,
    sheriffInPlayers,
  };
}

export { useGamePlayers };