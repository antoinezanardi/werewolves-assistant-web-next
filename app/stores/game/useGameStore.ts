import type { AsyncDataRequestStatus } from "nuxt/app";
import { defineStore } from "pinia";
import { unique } from "radash";

import type { MakeGamePlayDto } from "~/composables/api/game/dto/make-game-play/make-game-play.dto";
import type { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameEventsStore } from "~/stores/game/game-event/useGameEventsStore";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const {
    getGame: fetchGameFromApi,
    cancelGame: cancelGameFromApi,
    makeGamePlay: makeGamePlayFromApi,
  } = useFetchGames();

  const game = ref<Game>(new Game());
  const gameOptions = computed<GameOptions>(() => game.value.options);

  const gamePlayerGroups = computed<string[]>(() => unique(game.value.players.map(player => player.group)).filter(Boolean) as string[]);

  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const cancelingGameStatus = ref<AsyncDataRequestStatus>("idle");
  const makingGamePlayStatus = ref<AsyncDataRequestStatus>("idle");

  const { resetGameEventIndex } = useGameEventsStore();

  function resetGame(): void {
    game.value = new Game();
  }

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }

  async function cancelGame(): Promise<void> {
    cancelingGameStatus.value = "pending";
    const canceledGame = await cancelGameFromApi(game.value._id);
    if (!canceledGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = canceledGame;
    cancelingGameStatus.value = "success";
  }

  async function makeGamePlay(makeGamePlayDto: MakeGamePlayDto): Promise<void> {
    makingGamePlayStatus.value = "pending";
    const playedGame = await makeGamePlayFromApi(game.value._id, makeGamePlayDto);
    if (!playedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    resetGameEventIndex();
    game.value = playedGame;
    makingGamePlayStatus.value = "success";
  }

  async function skipGamePlay(): Promise<void> {
    return makeGamePlay({});
  }
  return {
    game,
    gameOptions,
    gamePlayerGroups,
    fetchingGameStatus,
    cancelingGameStatus,
    makingGamePlayStatus,
    resetGame,
    fetchAndSetGame,
    cancelGame,
    makeGamePlay,
    skipGamePlay,
  };
});

export { useGameStore };