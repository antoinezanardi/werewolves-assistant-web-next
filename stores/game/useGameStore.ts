import type { AsyncDataRequestStatus } from "#app/composables/asyncData";
import { defineStore } from "pinia";

import { Game } from "~/composables/api/game/types/game.class";
import { useFetchGames } from "~/composables/api/game/useFetchGames";
import { StoreIds } from "~/stores/enums/store.enum";

const useGameStore = defineStore(StoreIds.GAME, () => {
  const { getGame: fetchGameFromApi } = useFetchGames();

  const game = ref<Game>(new Game());
  const fetchingGameStatus = ref<AsyncDataRequestStatus>("idle");

  async function fetchAndSetGame(gameId: string): Promise<void> {
    fetchingGameStatus.value = "pending";
    const fetchedGame = await fetchGameFromApi(gameId);
    if (!fetchedGame) {
      fetchingGameStatus.value = "error";

      return;
    }
    game.value = fetchedGame;
    fetchingGameStatus.value = "success";
  }
  return {
    game,
    fetchingGameStatus,
    fetchAndSetGame,
  };
});

export { useGameStore };