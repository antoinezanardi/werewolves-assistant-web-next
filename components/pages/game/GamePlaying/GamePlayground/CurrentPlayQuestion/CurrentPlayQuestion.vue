<template>
  <h3
    id="current-play-question"
    class="text-center"
  >
    {{ currentPlayQuestion }}
  </h3>
</template>

<script setup lang="ts">
import { storeToRefs } from "pinia";

import type { GamePlaySourceName } from "~/composables/api/game/types/game-play/game-play-source/game-play-source.types";
import type { GamePlayAction, GamePlayCause } from "~/composables/api/game/types/game-play/game-play.types";
import { useGameStore } from "~/stores/game/useGameStore";

type CurrentPlayCausesQuestion = {
  [Key in GamePlayCause]?: string;
} & { default: string };

const gameStore = useGameStore();
const { game } = storeToRefs(gameStore);

const { t } = useI18n();

const gamePlaysQuestions: Record<GamePlaySourceName, Partial<Record<GamePlayAction, CurrentPlayCausesQuestion>>> = {
  "accursed-wolf-father": { infect: { default: t("components.CurrentPlayQuestion.doesAccursedWolfFatherWantToInfect") } },
  "actor": { "choose-card": { default: t("components.CurrentPlayQuestion.doesActorWantToChooseCard") } },
  "bear-tamer": { growl: { default: t("components.CurrentPlayQuestion.bearGrowled") } },
  "big-bad-wolf": { eat: { default: t("components.CurrentPlayQuestion.whichPlayerDoesBigBadWolfEat") } },
  "charmed": { "meet-each-other": { default: t("components.CurrentPlayQuestion.charmedPeopleMeetEachOther") } },
  "cupid": { charm: { default: t("components.CurrentPlayQuestion.whichPlayersDoesCupidCharm") } },
  "defender": { protect: { default: t("components.CurrentPlayQuestion.whichPlayerDoesDefenderProtect") } },
  "fox": { sniff: { default: t("components.CurrentPlayQuestion.doesFoxWantToSniff") } },
  "hunter": { shoot: { default: t("components.CurrentPlayQuestion.whichPlayerDoesHunterShoot") } },
  "lovers": { "meet-each-other": { default: t("components.CurrentPlayQuestion.loversMeetEachOther") } },
  "pied-piper": { charm: { default: t("components.CurrentPlayQuestion.whichPlayersDoesPiedPiperCharm") } },
  "scandalmonger": { mark: { default: t("components.CurrentPlayQuestion.doesScandalmongerWantToMark") } },
  "scapegoat": { "ban-voting": { default: t("components.CurrentPlayQuestion.doesScapegoatWantToBanVoting") } },
  "seer": { look: { default: t("components.CurrentPlayQuestion.whichPlayerDoesSeerLook") } },
  "sheriff": {
    "delegate": { default: t("components.CurrentPlayQuestion.whichPlayerDoesSheriffDelegate") },
    "settle-votes": { default: t("components.CurrentPlayQuestion.whichPlayerDoesSheriffSettleVotes") },
  },
  "stuttering-judge": { "request-another-vote": { default: t("components.CurrentPlayQuestion.doesJudgeRequestAnotherVote") } },
  "survivors": {
    "bury-dead-bodies": { default: "" },
    "elect-sheriff": { default: t("components.CurrentPlayQuestion.whichPlayerDoesSurvivorsElectAsSheriff") },
    "vote": {
      "default": t("components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteFor"),
      "angel-presence": t("components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteFor"),
      "stuttering-judge-request": t("components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteForOnJudgeRequest"),
      "previous-votes-were-in-ties": t("components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteForOnTie"),
    },
  },
  "thief": { "choose-card": { default: t("components.CurrentPlayQuestion.doesThiefWantToChooseCard") } },
  "three-brothers": { "meet-each-other": { default: t("components.CurrentPlayQuestion.threeBrothersMeetEachOther") } },
  "two-sisters": { "meet-each-other": { default: t("components.CurrentPlayQuestion.twoSistersMeetEachOther") } },
  "werewolves": { eat: { default: t("components.CurrentPlayQuestion.whichPlayerDoesWerewolvesEat") } },
  "white-werewolf": { eat: { default: t("components.CurrentPlayQuestion.doesWhiteWerewolfWantToEat") } },
  "wild-child": { "choose-model": { default: t("components.CurrentPlayQuestion.whichPlayerDoesWildChildChoose") } },
  "witch": { "use-potions": { default: t("components.CurrentPlayQuestion.doesWitchWantToUsePotions") } },
  "wolf-hound": { "choose-side": { default: t("components.CurrentPlayQuestion.whichSideDoesWolfHoundChoose") } },
};

const currentPlayQuestion = computed<string>(() => {
  const { currentPlay } = game.value;
  if (!currentPlay) {
    return t("components.CurrentPlayQuestion.unknownQuestionForCurrentPlay");
  }

  const { source, action, cause } = currentPlay;
  const question = gamePlaysQuestions[source.name][action]?.[cause ?? "default"];

  return question ?? t("components.CurrentPlayQuestion.unknownQuestionForCurrentPlay");
});
</script>