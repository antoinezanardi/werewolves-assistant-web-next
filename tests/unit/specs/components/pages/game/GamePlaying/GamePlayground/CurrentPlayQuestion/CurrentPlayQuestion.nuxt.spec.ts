import { createTestingPinia } from "@pinia/testing";
import type { mount } from "@vue/test-utils";

import CurrentPlayQuestion from "~/components/pages/game/GamePlaying/GamePlayground/CurrentPlayQuestion/CurrentPlayQuestion.vue";
import type { GamePlay } from "~/composables/api/game/types/game-play/game-play.class";
import { StoreIds } from "~/stores/enums/store.enum";
import { useGameStore } from "~/stores/game/useGameStore";
import { createFakeGamePlayAccursedWolfFatherInfects, createFakeGamePlayActorChoosesCard, createFakeGamePlayBearTamerGrowls, createFakeGamePlayBigBadWolfEats, createFakeGamePlayCharmedMeetEachOther, createFakeGamePlayCupidCharms, createFakeGamePlayDefenderProtects, createFakeGamePlayFoxSniffs, createFakeGamePlayHunterShoots, createFakeGamePlayLoversMeetEachOther, createFakeGamePlayPiedPiperCharms, createFakeGamePlayScandalmongerMarks, createFakeGamePlayScapegoatBansVoting, createFakeGamePlaySeerLooks, createFakeGamePlaySheriffDelegates, createFakeGamePlaySheriffSettlesVotes, createFakeGamePlayStutteringJudgeRequestsAnotherVote, createFakeGamePlaySurvivorsBuryDeadBodies, createFakeGamePlaySurvivorsElectSheriff, createFakeGamePlaySurvivorsVote, createFakeGamePlayThiefChoosesCard, createFakeGamePlayThreeBrothersMeetEachOther, createFakeGamePlayTwoSistersMeetEachOther, createFakeGamePlayWerewolvesEat, createFakeGamePlayWhiteWerewolfEats, createFakeGamePlayWildChildChoosesModel, createFakeGamePlayWitchUsesPotions, createFakeGamePlayWolfHoundChoosesSide } from "~/tests/unit/utils/factories/composables/api/game/game-play/game-play.factory";
import { createFakeGame } from "~/tests/unit/utils/factories/composables/api/game/game.factory";
import { mountSuspendedComponent } from "~/tests/unit/utils/helpers/mount.helpers";

describe("Current Play Question Component", () => {
  let wrapper: ReturnType<typeof mount<typeof CurrentPlayQuestion>>;
  const testingPinia = { initialState: { [StoreIds.GAME]: { game: createFakeGame({ currentPlay: createFakeGamePlayCupidCharms() }) } } };

  async function mountCurrentPlayQuestionComponent(): Promise<ReturnType<typeof mount<typeof CurrentPlayQuestion>>> {
    return mountSuspendedComponent(CurrentPlayQuestion, { global: { plugins: [createTestingPinia(testingPinia)] } });
  }

  beforeEach(async() => {
    wrapper = await mountCurrentPlayQuestionComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it.each<{
    currentPlay: GamePlay | null;
    expectedQuestion: string;
    test: string;
  }>([
    {
      currentPlay: null,
      expectedQuestion: "components.CurrentPlayQuestion.unknownQuestionForCurrentPlay",
      test: "should render nothing when current play is null.",
    },
    {
      currentPlay: createFakeGamePlayAccursedWolfFatherInfects(),
      expectedQuestion: "components.CurrentPlayQuestion.doesAccursedWolfFatherWantToInfect",
      test: "should render the question for accursed wolf father infects when current play is accursed wolf father who infects.",
    },
    {
      currentPlay: createFakeGamePlayActorChoosesCard(),
      expectedQuestion: "components.CurrentPlayQuestion.doesActorWantToChooseCard",
      test: "should render the question for actor chooses card when current play is actor who chooses card.",
    },
    {
      currentPlay: createFakeGamePlayBearTamerGrowls(),
      expectedQuestion: "components.CurrentPlayQuestion.bearGrowled",
      test: "should render the question for bear tamer growls when current play is bear tamer who growls.",
    },
    {
      currentPlay: createFakeGamePlayBigBadWolfEats(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesBigBadWolfEat",
      test: "should render the question for big bad wolf eats when current play is big bad wolf who eats.",
    },
    {
      currentPlay: createFakeGamePlayCharmedMeetEachOther(),
      expectedQuestion: "components.CurrentPlayQuestion.charmedPeopleMeetEachOther",
      test: "should render the question for charmed meet each other when current play is charmed who meet each other.",
    },
    {
      currentPlay: createFakeGamePlayCupidCharms(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayersDoesCupidCharm",
      test: "should render the question for cupid charms when current play is cupid who charms.",
    },
    {
      currentPlay: createFakeGamePlayDefenderProtects(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesDefenderProtect",
      test: "should render the question for defender protects when current play is defender who protects.",
    },
    {
      currentPlay: createFakeGamePlayFoxSniffs(),
      expectedQuestion: "components.CurrentPlayQuestion.doesFoxWantToSniff",
      test: "should render the question for fox sniffs when current play is fox who sniffs.",
    },
    {
      currentPlay: createFakeGamePlayHunterShoots(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesHunterShoot",
      test: "should render the question for hunter shoots when current play is hunter who shoots.",
    },
    {
      currentPlay: createFakeGamePlayLoversMeetEachOther(),
      expectedQuestion: "components.CurrentPlayQuestion.loversMeetEachOther",
      test: "should render the question for lovers meet each other when current play is lovers who meet each other.",
    },
    {
      currentPlay: createFakeGamePlayPiedPiperCharms(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayersDoesPiedPiperCharm",
      test: "should render the question for pied piper charms when current play is pied piper who charms.",
    },
    {
      currentPlay: createFakeGamePlayScandalmongerMarks(),
      expectedQuestion: "components.CurrentPlayQuestion.doesScandalmongerWantToMark",
      test: "should render the question for scandalmonger marks when current play is scandalmonger who marks.",
    },
    {
      currentPlay: createFakeGamePlayScapegoatBansVoting(),
      expectedQuestion: "components.CurrentPlayQuestion.doesScapegoatWantToBanVoting",
      test: "should render the question for scapegoat bans voting when current play is scapegoat who bans voting.",
    },
    {
      currentPlay: createFakeGamePlaySeerLooks(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSeerLook",
      test: "should render the question for seer looks when current play is seer who looks.",
    },
    {
      currentPlay: createFakeGamePlaySheriffDelegates(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSheriffDelegate",
      test: "should render the question for sheriff delegates when current play is sheriff who delegates.",
    },
    {
      currentPlay: createFakeGamePlaySheriffSettlesVotes(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSheriffSettleVotes",
      test: "should render the question for sheriff settles votes when current play is sheriff who settles votes.",
    },
    {
      currentPlay: createFakeGamePlayStutteringJudgeRequestsAnotherVote(),
      expectedQuestion: "components.CurrentPlayQuestion.doesJudgeRequestAnotherVote",
      test: "should render the question for stuttering judge requests another vote when current play is stuttering judge who requests another vote.",
    },
    {
      currentPlay: createFakeGamePlaySurvivorsBuryDeadBodies(),
      expectedQuestion: "",
      test: "should render nothing when current play is survivors bury dead bodies.",
    },
    {
      currentPlay: createFakeGamePlaySurvivorsElectSheriff(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSurvivorsElectAsSheriff",
      test: "should render the question for survivors elect sheriff when current play is survivors who elect sheriff.",
    },
    {
      currentPlay: createFakeGamePlaySurvivorsVote(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteFor",
      test: "should render the question for survivors vote when current play is survivors who vote and there is no cause.",
    },
    {
      currentPlay: createFakeGamePlaySurvivorsVote({ cause: "angel-presence" }),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteFor",
      test: "should render the question for survivors vote when current play is survivors who vote and cause is angel presence.",
    },
    {
      currentPlay: createFakeGamePlaySurvivorsVote({ cause: "stuttering-judge-request" }),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteForOnJudgeRequest",
      test: "should render the question for survivors vote when current play is survivors who vote and cause is stuttering judge request.",
    },
    {
      currentPlay: createFakeGamePlaySurvivorsVote({ cause: "previous-votes-were-in-ties" }),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesSurvivorsVoteForOnTie",
      test: "should render the question for survivors vote when current play is survivors who vote and cause is stuttering judge request and target is player 1.",
    },
    {
      currentPlay: createFakeGamePlayThiefChoosesCard(),
      expectedQuestion: "components.CurrentPlayQuestion.doesThiefWantToChooseCard",
      test: "should render the question for thief chooses card when current play is thief who chooses card.",
    },
    {
      currentPlay: createFakeGamePlayThreeBrothersMeetEachOther(),
      expectedQuestion: "components.CurrentPlayQuestion.threeBrothersMeetEachOther",
      test: "should render the question for three brothers meet each other when current play is three brothers who meet each other.",
    },
    {
      currentPlay: createFakeGamePlayTwoSistersMeetEachOther(),
      expectedQuestion: "components.CurrentPlayQuestion.twoSistersMeetEachOther",
      test: "should render the question for two sisters meet each other when current play is two sisters who meet each other.",
    },
    {
      currentPlay: createFakeGamePlayWerewolvesEat(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesWerewolvesEat",
      test: "should render the question for werewolves eat when current play is werewolves who eat.",
    },
    {
      currentPlay: createFakeGamePlayWhiteWerewolfEats(),
      expectedQuestion: "components.CurrentPlayQuestion.doesWhiteWerewolfWantToEat",
      test: "should render the question for white werewolf eats when current play is white werewolf who eats.",
    },
    {
      currentPlay: createFakeGamePlayWildChildChoosesModel(),
      expectedQuestion: "components.CurrentPlayQuestion.whichPlayerDoesWildChildChoose",
      test: "should render the question for wild child chooses model when current play is wild child who chooses model.",
    },
    {
      currentPlay: createFakeGamePlayWitchUsesPotions(),
      expectedQuestion: "components.CurrentPlayQuestion.doesWitchWantToUsePotions",
      test: "should render the question for witch uses potions when current play is witch who uses potions.",
    },
    {
      currentPlay: createFakeGamePlayWolfHoundChoosesSide(),
      expectedQuestion: "components.CurrentPlayQuestion.whichSideDoesWolfHoundChoose",
      test: "should render the question for wolf hound chooses side when current play is wolf hound who chooses side.",
    },
    {
      currentPlay: createFakeGamePlayWitchUsesPotions({ action: "shoot" }),
      expectedQuestion: "components.CurrentPlayQuestion.unknownQuestionForCurrentPlay",
      test: "should render unknown question when action is not relevant.",
    },
    {
      currentPlay: createFakeGamePlayWitchUsesPotions({ cause: "stuttering-judge-request" }),
      expectedQuestion: "components.CurrentPlayQuestion.unknownQuestionForCurrentPlay",
      test: "should render unknown question when cause is not relevant.",
    },
  ])("$test", async({ currentPlay, expectedQuestion }) => {
    const gameStore = useGameStore();
    gameStore.game.currentPlay = currentPlay;
    await nextTick();
    const question = wrapper.find<HTMLHeadingElement>("#current-play-question");

    expect(question.text()).toBe(expectedQuestion);
  });
});