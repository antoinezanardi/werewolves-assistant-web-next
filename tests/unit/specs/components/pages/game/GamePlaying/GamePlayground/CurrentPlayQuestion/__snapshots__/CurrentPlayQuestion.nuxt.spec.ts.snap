// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Current Play Question Component > should match snapshot when rendered. 1`] = `"<h3 id="current-play-question" class="px-4 text-center">components.CurrentPlayQuestion.whichPlayersDoesCupidCharm</h3>"`;
